File: smtk/attribute/Testing/basicAttributeDerivationTest.cxx
/*=========================================================================

Copyright (c) 1998-2012 Kitware Inc. 28 Corporate Drive,
Clifton Park, NY, 12065, USA.

All rights reserved. No part of this software may be reproduced, distributed,
or modified, in any form or by any means, without permission in writing from
Kitware Inc.

IN NO EVENT SHALL THE AUTHORS OR DISTRIBUTORS BE LIABLE TO ANY PARTY FOR
DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT
OF THE USE OF THIS SOFTWARE, ITS DOCUMENTATION, OR ANY DERIVATIVES THEREOF,
EVEN IF THE AUTHORS HAVE BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

THE AUTHORS AND DISTRIBUTORS SPECIFICALLY DISCLAIM ANY WARRANTIES,
INCLUDING,
BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE IS PROVIDED ON AN
"AS IS" BASIS, AND THE AUTHORS AND DISTRIBUTORS HAVE NO OBLIGATION TO
PROVIDE
MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
=========================================================================*/

#include "smtk/attribute/Manager.h"
#include "smtk/attribute/Definition.h"
#include "smtk/attribute/Attribute.h"
#include "smtk/attribute/IntItem.h"
#include "smtk/attribute/IntItemDefinition.h"
#include "smtk/attribute/DoubleItem.h"
#include "smtk/attribute/DoubleItemDefinition.h"
#include "smtk/attribute/StringItem.h"
#include "smtk/attribute/StringItemDefinition.h"
#include <iostream>

int main()
{
  int status = 0;
  {
  typedef smtk::attribute::IntItemDefinition IntCompDef;
  typedef smtk::attribute::DoubleItemDefinition DoubleCompDef;
  typedef smtk::attribute::StringItemDefinition StringCompDef;
  typedef smtk::attribute::ValueItem ValueComp;
  typedef smtk::attribute::Item AttComp;

  smtk::attribute::Manager manager;
  std::cout << "Manager Created\n";
  smtk::attribute::DefinitionPtr base = manager.createDefinition("BaseDef");
  // Lets add some item definitions
  smtk::attribute::IntItemDefinitionPtr icompdef = IntCompDef::New("IntComp1");
  base->addItemDefinition(icompdef);
  smtk::attribute::IntItemDefinitionPtr icompdef2 = IntCompDef::New("IntComp2");
  icompdef2->setDefaultValue(10);
  base->addItemDefinition(icompdef2);

  smtk::attribute::DefinitionPtr def1 = manager.createDefinition("Derived1", "BaseDef");
   // Lets add some item definitions
  smtk::attribute::DoubleItemDefinitionPtr dcompdef = DoubleCompDef::New("DoubleComp1");
  def1->addItemDefinition(dcompdef);
  smtk::attribute::DoubleItemDefinitionPtr dcompdef2 = DoubleCompDef::New("DoubleComp2");
  dcompdef2->setDefaultValue(-35.2);
  def1->addItemDefinition(dcompdef2);

  smtk::attribute::DefinitionPtr def2 = manager.createDefinition("Derived2", "Derived1");
   // Lets add some item definitions
  smtk::attribute::StringItemDefinitionPtr scompdef = StringCompDef::New("StringComp1");
  def1->addItemDefinition(scompdef);
  smtk::attribute::StringItemDefinitionPtr scompdef2 = StringCompDef::New("StringComp2");
  scompdef2->setDefaultValue("Default");
  def1->addItemDefinition(scompdef2);

  smtk::attribute::AttributePtr att = manager.createAttribute("testAtt", "Derived2");
  if (att)
    {
    std::cout << "Attribute testAtt created\n";
    }
  else
    {
    std::cout << "ERROR: Attribute testAtt not created\n";
    status = -1;
    }

  smtk::attribute::ValueItemPtr vcomp;
  smtk::attribute::ItemPtr comp;

  int i, n = static_cast<int>(att->numberOfItems());
  std::cout << "Items of testAtt:\n";
  for (i = 0; i < n; i++)
    {
    comp = att->item(i);
    std::cout << "\t" << comp->name() << " Type = " << AttComp::type2String(comp->type()) << ", ";
    vcomp = smtk::dynamic_pointer_cast<ValueComp>(comp);
    if (vcomp)
      {
      switch (vcomp->type())
        {
        case AttComp::DOUBLE:
        case AttComp::INT:
          std::cout << " Value = "  << vcomp->valueAsString() << std::endl;
          break;
        case AttComp::STRING:
          std::cout << " String Val = " << vcomp->valueAsString() << std::endl;
          break;
        default:
          break;
        }
      }
    }
  std::cout << "Manager destroyed\n";
  }
  return status;
}

File: smtk/attribute/testing/python/basicAttributeDerivationTest.py
#=============================================================================
#
#  Copyright (c) Kitware, Inc.
#  All rights reserved.
#  See LICENSE.txt for details.
#
#  This software is distributed WITHOUT ANY WARRANTY; without even
#  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
#  PURPOSE.  See the above copyright notice for more information.
#
#=============================================================================
"""
Manual port of SMTK/smtk/attribute/Testing/basicAttributeDerivationTest.cxx
For verifying python-shiboken wrappers

Requires SMTKCorePython.so to be in module path
"""

import smtk

itemNames = ["IntComp1", "IntComp2", "DoubleComp1",
             "DoubleComp2", "StringComp1", "StringComp2",
             "ModelComp1_FACE", "ModelComp2_GROUP"]

if __name__ == '__main__':
    import sys

    status = 0

    system = smtk.attribute.System()
    print 'System created'
    base = system.createDefinition("BaseDef");

    #Lets add some item definitions
    icompdef = smtk.attribute.IntItemDefinition.New(itemNames[0])
    itemdef = smtk.attribute.IntItemDefinition.ToItemDefinition(icompdef)
    base.addItemDefinition(itemdef);
    icompdef2 = smtk.attribute.IntItemDefinition.New(itemNames[1])
    icompdef2.setDefaultValue(10);
    itemdef2 = smtk.attribute.IntItemDefinition.ToItemDefinition(icompdef2)
    base.addItemDefinition(itemdef2);

    def1 = system.createDefinition('Derived1', 'BaseDef')
    # Lets add some item definitions
    dcompdef = smtk.attribute.DoubleItemDefinition.New(itemNames[2])
    itemdef = smtk.attribute.DoubleItemDefinition.ToItemDefinition(dcompdef)
    def1.addItemDefinition(itemdef)
    dcompdef2 = smtk.attribute.DoubleItemDefinition.New(itemNames[3])
    dcompdef2.setDefaultValue(-35.2)
    itemdef2 = smtk.attribute.DoubleItemDefinition.ToItemDefinition(dcompdef2)
    def1.addItemDefinition(itemdef2)

    def2 = system.createDefinition('Derived2', 'Derived1')
    # Lets add some item definitions
    scompdef = smtk.attribute.StringItemDefinition.New(itemNames[4])
    itemdef = smtk.attribute.StringItemDefinition.ToItemDefinition(scompdef)
    def2.addItemDefinition(itemdef)
    scompdef2 = smtk.attribute.StringItemDefinition.New(itemNames[5])
    scompdef2.setDefaultValue('Default')
    itemdef2 = smtk.attribute.StringItemDefinition.ToItemDefinition(scompdef2)
    def2.addItemDefinition(itemdef2)

    def3 = system.createDefinition('Derived3', 'Derived1')
    # Lets add some item definitions
    scompdef = smtk.attribute.ModelEntityItemDefinition.New(itemNames[6])
    scompdef.setMembershipMask(smtk.model.FACE)
    scompdef.setIsExtensible(True) # Need to set this or numberOfRequiredValues
    itemdef = smtk.attribute.ModelEntityItemDefinition.ToItemDefinition(scompdef)
    def3.addItemDefinition(itemdef)
    scompdef3 = smtk.attribute.ModelEntityItemDefinition.New(itemNames[7])
    scompdef3.setMembershipMask(smtk.model.CELL_ENTITY | smtk.model.GROUP_ENTITY | smtk.model.HOMOGENOUS_GROUP)
    scompdef3.setIsExtensible(True) # Need to set this or numberOfRequiredValues
    itemdef3 = smtk.attribute.ModelEntityItemDefinition.ToItemDefinition(scompdef3)
    def3.addItemDefinition(itemdef3)

    # Lets test out the find item position method
    pstatus = 0;
    for j in range(0,6):
      if (def2.findItemPosition(itemNames[j]) != j):
        print "Incorrect Position Returned for %s, position returned is %d, but it should be %d\n" % (itemNames[j], def2.findItemPosition(itemNames[j]), j)
        pstatus = -1
    if pstatus == 0:
      print "Initial Position Test Passed!\n"
    else:
      print "Initial Position Test Failed!\n"
      status = -1

    # Lets add a  component to the base def and verify that positions are reordered
    base.addItemDefinitionStr(smtk.attribute.StringItemDefinition, "InsertStringItem")
    pstatus = 0
    for j in range(2,6):
      if def2.findItemPosition(itemNames[j]) != (j+1):
        print "Incorrect Position Returned for %s, position returned is %d, but it should be %d\n" % (itemNames[j], def2.findItemPosition(itemNames[j]), j+1)
        pstatus = -1;

    if pstatus == 0:
      print "Insertion Position Test Passed!\n"
    else:
      print "Insertion Position Test Failed!\n"
      status = -1

    att = system.createAttribute('testAtt', 'Derived2')
    if not att is None:
        print 'Attribute testAtt created'
    else:
        print 'ERROR: Attribute testAtt not created'
        status = -1
    comp = att.find("DoubleComp1")
    if not comp is None:
       vvcomp = smtk.attribute.ValueItem.CastTo(comp)
       print " Value = %s"%vvcomp.valueAsString()
    else:
       print "ERROR: could not find the base's item"
       status = -1;

    n = att.numberOfItems()
    for i in range(n):
        comp = att.item(i)
        print '\t%s Type = %s, ' % (comp.name(), \
          smtk.attribute.Item.type2String(comp.type())),
        vcomp = smtk.attribute.ValueItem.CastTo(comp)
        if vcomp:
          if vcomp.type() == smtk.attribute.Item.DOUBLE:
            print ' Value = %s' % vcomp.valueAsString()
          elif vcomp.type() == smtk.attribute.Item.INT:
            print ' Value = %s' % vcomp.valueAsString()
          elif vcomp.type() == smtk.attribute.Item.STRING:
            print 'String Val = %s' % vcomp.valueAsString()
        else:
          ecomp = smtk.attribute.ModelEntityItem.CastTo(comp)
          if ecomp:
            if ecomp.type() == smtk.attribute.Item.MODEL_ENTITY:
              print ' Value = %s' % ecomp.valueAsString()
          else:
            print comp

    # ======
    # Now test setting entity-valued attribute-items
    # I. Create a model manager and add some entities to it:
    mmgr = smtk.model.Manager.create()
    system.setRefModelManager(mmgr)
    mdl = mmgr.addModel(3,3,'TestModel')
    edg = mmgr.addEdge()
    fac = mmgr.addFace()
    gr0 = mmgr.addGroup(smtk.model.HOMOGENOUS_GROUP, 'GroupA')
    gr1 = mmgr.addGroup(smtk.model.PARTITION, 'GroupB')
    gr0.addEntity(fac)
    gr1.addEntity(edg)

    # II. Have the attribute system create an attribute with model-entity
    att = system.createAttribute('testMEAtt', 'Derived3')
    if not att is None:
        print 'Attribute testMEAtt created'
    else:
        print 'ERROR: Attribute testMEAtt not created'
        status = -1
    # III. Find the face-only attribute item and try fuzzing its arguments a bit
    comp = att.find("ModelComp1_FACE")
    if not comp is None:
       fcomp = smtk.attribute.ModelEntityItem.CastTo(comp)
       if fcomp.appendValue(edg):
         print 'ERROR: Face-only attribute value accepted an edge'
         status = -1
       if not fcomp.appendValue(fac):
         print 'ERROR: Face-only attribute value did not accept a face'
         status = -1
       print " Value = %s" % fcomp.valueAsString()
    else:
       print "ERROR: could not find the base's item"
       status = -1;
    # IV. Find the homogenous-group-only attribute item and try fuzzing its arguments a bit
    comp = att.find("ModelComp2_GROUP")
    if not comp is None:
       gcomp = smtk.attribute.ModelEntityItem.CastTo(comp)
       if not gcomp.appendValue(gr0):
         print 'ERROR: Homogenous-group-only attribute value did not accept a homogenous group'
         status = -1
       if gcomp.appendValue(gr1):
         print 'ERROR: Homogenous-group-only attribute value accepted a more relaxed group'
         status = -1
       print " Value = %s" % gcomp.valueAsString()
    else:
       print "ERROR: could not find the base's item"
       status = -1;

    n = att.numberOfItems()
    for i in range(n):
        comp = att.item(i)
        print '\t%s Type = %s, ' % (comp.name(), \
          smtk.attribute.Item.type2String(comp.type())),
        vcomp = smtk.attribute.ValueItem.CastTo(comp)
        if vcomp:
          if vcomp.type() == smtk.attribute.Item.DOUBLE:
            print ' Value = %s' % vcomp.valueAsString()
          elif vcomp.type() == smtk.attribute.Item.INT:
            print ' Value = %s' % vcomp.valueAsString()
          elif vcomp.type() == smtk.attribute.Item.STRING:
            print 'String Val = %s' % vcomp.valueAsString()
        else:
          ecomp = smtk.attribute.ModelEntityItem.CastTo(comp)
          if ecomp:
            if ecomp.type() == smtk.attribute.Item.MODEL_ENTITY:
              print ' Value = %s' % ecomp.valueAsString()
          else:
            print comp
    # ======

    del system
    print 'System destroyed'

    sys.exit(status)
